File: vscode-extension/src/antares/agent/toolExecution.ts
// Port of antares_cli/agent/tool_execution.py — budget, dispatch, output formatting.

import { MAX_TOOL_OUTPUT_CHARS } from "../sandbox/shellPolicy";
import { ToolCallResult } from "./modelAdapter";
import { validateToolCallSafety } from "./quarantine";
import { ModelSessionState } from "./state";
import { ToolRouter } from "./toolRouter";

const REPOSITORY_COMMAND_TOOLS = new Set(["bash", "terminal", "read_file"]);
const MAX_TOOL_ERROR_CHARS = 2_000;
const TOOL_ERROR_TRUNCATION_SUFFIX = "...[tool error truncated]";
const AVAILABLE_TOOLS = "terminal, read_file, submit_vulnerable_files, submit_no_vulnerability_found";

export class AgentToolExecutor {
  constructor(private readonly toolRouter: ToolRouter) {}

  execute(
    toolName: string,
    args: Record<string, unknown>,
    state: ModelSessionState
  ): ToolCallResult {
    state.toolCallCount += 1;
    state.sessionTrace.recordToolCall({ toolName, arguments: args });

    const normalized = toolName.trim().toLowerCase();
    if (REPOSITORY_COMMAND_TOOLS.has(normalized)) {
      const budget = state.terminalCallBudget;
      if (budget !== null && state.terminalCallsUsed >= budget) {
        state.failedToolCallsCount += 1;
        return {
          toolName,
          toolResponse: `Terminal call budget exhausted (${budget}/${budget}). Submit your answer.`,
        };
      }
      state.terminalCallsUsed += 1;
    }

    const safety = validateToolCallSafety(toolName, args);
    if (safety.blocked) {
      state.failedToolCallsCount += 1;
      return { toolName, toolResponse: `Tool call blocked: ${safety.reason}` };
    }

    const result = this.toolRouter.execute(toolName, args);
    let response: string;
    if (!result.success) {
      response = this.recordToolError(toolName, result.errorMessage ?? "unknown error", state);
    } else {
      state.consecutiveErrors = 0;
      response = formatToolOutput(result.output);
    }

    const sanitized = state.contentQuarantine.sanitize(response);
    state.trajectory.push({ entry_type: "tool_call", content: `→ ${toolName}(${argsPreview(args)})` });
    state.trajectory.push({ entry_type: "tool_response", content: sanitized.slice(0, 500) });
    return { toolName, toolResponse: sanitized };
  }

  private recordToolError(toolName: string, error: string, state: ModelSessionState): string {
    state.consecutiveErrors += 1;
    state.failedToolCallsCount += 1;
    let message = boundToolError(`Tool ${toolName} failed: ${error}`);
    if (state.consecutiveErrors >= 2) {
      message += `\n\nAvailable tools: ${AVAILABLE_TOOLS}. Use one of these exact names.`;
    }
    return message;
  }
}

function argsPreview(args: Record<string, unknown>): string {
  return Object.entries(args)
    .map(([key, value]) => `${key}=${typeof value === "string" ? value : JSON.stringify(value)}`)
    .join(", ");
}

function formatToolOutput(output: Record<string, unknown>): string {
  if ("stdout" in output) {
    let text = String(output.stdout ?? "");
    const stderr = output.stderr ? String(output.stderr) : "";
    if (stderr) {
      text += `\n[stderr]: ${stderr}`;
    }
    if (output.truncated) {
      text +=
        `\n\n[OUTPUT TRUNCATED: showing first ${MAX_TOOL_OUTPUT_CHARS.toLocaleString("en-US")} ` +
        "characters. Use head/tail/sed with line ranges to read specific sections.]";
    }
    return text;
  }
  return JSON.stringify(output);
}

function boundToolError(message: string): string {
  const sanitized = message.replace(/[^\P{C}\n\t]/gu, "");
  if (sanitized.length <= MAX_TOOL_ERROR_CHARS) {
    return sanitized;
  }
  return sanitized.slice(0, MAX_TOOL_ERROR_CHARS - TOOL_ERROR_TRUNCATION_SUFFIX.length) + TOOL_ERROR_TRUNCATION_SUFFIX;
}

File: src/antares_cli/agent/tool_execution.py
"""Tool execution side effects for model-driven audits."""

from __future__ import annotations

from antares_cli.agent.contracts import BASH_TOOL_NAME, TERMINAL_TOOL_NAME
from antares_cli.agent.quarantine import validate_tool_call_safety
from antares_cli.agent.state import ModelSessionState
from antares_cli.agent.streaming import ParsedToolCall
from antares_cli.agent.tool_router import ToolRouter, bound_tool_error
from antares_cli.output.finding import TrajectoryEntry
from antares_cli.tools.shell_exec import MAX_TOOL_OUTPUT_CHARS

REPOSITORY_COMMAND_TOOLS = {BASH_TOOL_NAME, TERMINAL_TOOL_NAME, "read_file"}


class AgentToolExecutor:
    """Executes model-requested tools and records all related session side effects."""

    def __init__(self, *, tool_router: ToolRouter) -> None:
        self.tool_router = tool_router

    def execute(self, parsed_call: ParsedToolCall, state: ModelSessionState) -> str:
        state.tool_call_count += 1
        self._record_tool_call(parsed_call, state)
        if parsed_call.tool_name in REPOSITORY_COMMAND_TOOLS:
            if self._terminal_budget_exhausted(state):
                state.failed_tool_calls_count += 1
                tool_response_text = self._record_budget_exhaustion(parsed_call.tool_name, state)
                self._record_trajectory(parsed_call, state, tool_response_text)
                return tool_response_text
            state.terminal_calls_used += 1

        safety_check = validate_tool_call_safety(
            parsed_call.tool_name,
            parsed_call.arguments,
            allowed_sensitive_files=self.tool_router.allowed_sensitive_files,
        )
        if not safety_check.safe:
            state.failed_tool_calls_count += 1
            tool_response_text = f"Tool call blocked: {safety_check.blocked_reason}"
            state.session_trace.record_event(
                phase="tool_blocked",
                payload={
                    "tool_name": parsed_call.tool_name,
                    "reason": safety_check.blocked_reason or "",
                },
            )
        else:
            tool_response_text = self._execute_safe_tool_call(parsed_call, state)

        self._record_trajectory(parsed_call, state, tool_response_text)
        return tool_response_text

    @staticmethod
    def _terminal_budget_exhausted(state: ModelSessionState) -> bool:
        budget = state.terminal_call_budget
        return budget is not None and state.terminal_calls_used >= budget

    @staticmethod
    def _record_budget_exhaustion(tool_name: str, state: ModelSessionState) -> str:
        budget = state.terminal_call_budget
        if budget is None:
            raise RuntimeError(
                "Repository tool budget exhaustion was recorded without a configured budget"
            )
        message = f"Terminal call budget exhausted ({budget}/{budget}). Submit your answer."
        state.session_trace.record_event(
            phase="tool_blocked",
            payload={"tool_name": tool_name, "reason": message},
        )
        state.session_trace.record_tool_result(
            tool_name=tool_name,
            result_summary=message,
            evidence_id="budget_exhausted",
        )
        return message

    def _record_tool_call(
        self,
        parsed_call: ParsedToolCall,
        state: ModelSessionState,
    ) -> None:
        state.session_trace.record_tool_call(
            tool_name=parsed_call.tool_name,
            arguments=parsed_call.arguments,
        )

    def _execute_safe_tool_call(
        self,
        parsed_call: ParsedToolCall,
        state: ModelSessionState,
    ) -> str:
        execution_result = self.tool_router.execute(
            parsed_call.tool_name,
            parsed_call.arguments,
        )
        if not execution_result.success:
            return self._record_tool_error(parsed_call, state, execution_result.error_message)
        state.consecutive_errors = 0
        tool_response_text = self._sanitize_tool_output(
            parsed_call,
            state,
            _format_tool_output(execution_result.output),
        )
        state.session_trace.record_tool_result(
            tool_name=parsed_call.tool_name,
            result_summary=tool_response_text[:500],
            evidence_id=state.session_trace.new_evidence_id(),
        )
        return tool_response_text

    def _record_tool_error(
        self,
        parsed_call: ParsedToolCall,
        state: ModelSessionState,
        error_message: str | None,
    ) -> str:
        state.failed_tool_calls_count += 1
        state.consecutive_errors += 1
        tool_response_text = bound_tool_error(
            f"Tool {parsed_call.tool_name} failed: {error_message}"
        )
        if state.consecutive_errors >= 2:
            available = ", ".join(self.tool_router.available_tool_names())
            tool_response_text = bound_tool_error(
                tool_response_text
                + f"\n\nAvailable tools: {available}. Use one of these exact names."
            )
        state.session_trace.record_tool_result(
            tool_name=parsed_call.tool_name,
            result_summary=tool_response_text,
            evidence_id="error",
        )
        return tool_response_text

    @staticmethod
    def _sanitize_tool_output(
        parsed_call: ParsedToolCall,
        state: ModelSessionState,
        raw_output: str,
    ) -> str:
        scan_result = state.content_quarantine.scan(raw_output)
        if not scan_result.injection_detected:
            return scan_result.sanitized_text
        alert_source = _alert_source(parsed_call)
        state.session_trace.record_event(
            phase="quarantine_alert",
            payload={"source": alert_source, "count": len(scan_result.quarantined_spans)},
        )
        return scan_result.sanitized_text

    @staticmethod
    def _record_trajectory(
        parsed_call: ParsedToolCall,
        state: ModelSessionState,
        tool_response_text: str,
    ) -> None:
        full_args = ", ".join(f"{key}={value}" for key, value in parsed_call.arguments.items())
        state.trajectory.append(
            TrajectoryEntry(
                entry_type="tool_call",
                content=f"→ {parsed_call.tool_name}({full_args})",
            )
        )
        state.trajectory.append(
            TrajectoryEntry(
                entry_type="tool_response",
                content=tool_response_text,
            )
        )


def _format_tool_output(output: object) -> str:
    """Format tool execution output for the model, surfacing truncation."""
    if not isinstance(output, dict):
        return str(output)
    stdout = output.get("stdout", "")
    stderr = output.get("stderr", "")
    truncated = output.get("truncated", False)
    result = str(stdout)
    if stderr:
        result += f"\n[stderr]: {stderr}"
    if truncated:
        result += (
            f"\n\n[OUTPUT TRUNCATED: showing first {MAX_TOOL_OUTPUT_CHARS:,} characters. "
            "Use head/tail/sed with line ranges to read specific sections.]"
        )
    return result


def _alert_source(parsed_call: ParsedToolCall) -> str:
    args_summary = ", ".join(
        f"{key}={str(value)[:30]}" for key, value in list(parsed_call.arguments.items())[:2]
    )
    return f"{parsed_call.tool_name}({args_summary})"
